### PR TITLE
[decoupled-execution] Integration Part 1 - Adapt safety-rules

### PR DESCRIPTION
**Pull request by yuxiamit**
This is the first part of integrating decoupled execution. We add
a config item named `decoupled_execution` for safety-rules.

## Motivation

(Write your motivation for proposed changes here.)

### Have you read the [Contributing Guidelines on pull requests](https://github.com/diem/diem/blob/main/CONTRIBUTING.md#pull-requests)?

(Write your answer here.)

## Test Plan

(Share your test plan here. If you changed code, please provide us with clear instructions for verifying that your changes work.)

## Related PRs

(If this PR adds or changes functionality, please take some time to update or suggest changes to the docs at https://developers.diem.com, and link to your PR here.)

## If targeting a release branch, please fill the below out as well

 * Justification and breaking nature (who does it affect? validators, full nodes, tooling, operators, AOS, etc.)
 * Comprehensive test results that demonstrate the fix working and not breaking existing workflows.
 * Why we must have it for V1 launch.
 * What workarounds and alternative we have if we do not push the PR.

### CODE REVIEW

File: consensus/safety-rules/src/safety_rules.rs [hunk 1]
```diff
@@ -57,13 +58,18 @@ impl SafetyRules {
...
         export_consensus_key: bool,
+        decoupled_execution: bool,
     ) -> Self {
         let execution_public_key = if verify_vote_proposal_signature {
-            Some(
-                persistent_storage
-                    .execution_public_key()
-                    .expect("Unable to retrieve execution public key"),
-            )
+            if decoupled_execution {
```
Comment (zekun000) [inline consensus/safety-rules/src/safety_rules.rs:64]: if v && !d

File: consensus/safety-rules/src/safety_rules.rs [hunk 2]
```diff
@@ -85,10 +92,12 @@ impl SafetyRules {
         let execution_signature = maybe_signed_vote_proposal.signature.as_ref();
 
         if let Some(public_key) = self.execution_public_key.as_ref() {
-            execution_signature
-                .ok_or(Error::VoteProposalSignatureNotFound)?
-                .verify(vote_proposal, public_key)
-                .map_err(|error| Error::InternalError(error.to_string()))?;
+            if self.decoupled_execution {
```
Comment (zekun000) [inline consensus/safety-rules/src/safety_rules.rs:95]: if !self.decoupled_execution?

File: consensus/safety-rules/src/safety_rules.rs [hunk 3]
```diff
@@ -103,7 +112,18 @@ impl SafetyRules {
...
+            Ok(VoteData::new(
+                proposed_block.gen_block_info(
+                    *ACCUMULATOR_PLACEHOLDER_HASH,
+                    0,
+                    vote_proposal.next_epoch_state().cloned(),
+                ),
+                proposed_block.quorum_cert().certified_block().clone(),
+            ))
+        } else {
+            self.extension_check(vote_proposal)
```
Comment (davidiw) [inline consensus/safety-rules/src/safety_rules.rs:125]: feels like we could add two methods to VoteProposal, vote_data and vote_data_with_extension_proof and eliminate this code from here?

File: consensus/safety-rules/src/safety_rules.rs (hunk 1 above)
Comment (yuxiamit) [reply consensus/safety-rules/src/safety_rules.rs:64]: resolved

File: consensus/safety-rules/src/safety_rules.rs (hunk 2 above)
Comment (yuxiamit) [reply consensus/safety-rules/src/safety_rules.rs:95]: nice catch!

File: consensus/safety-rules/src/safety_rules.rs (hunk 3 above)
Comment (yuxiamit) [reply consensus/safety-rules/src/safety_rules.rs:125]: Thanks! This makes the code cleaner.

File: consensus/consensus-types/src/vote_proposal.rs [hunk 4]
```diff
@@ -51,6 +57,31 @@ impl VoteProposal {
     pub fn next_epoch_state(&self) -> Option<&EpochState> {
         self.next_epoch_state.as_ref()
     }
+
+    pub fn vote_data(&self) -> VoteData {
```
Comment (zekun000) [inline consensus/consensus-types/src/vote_proposal.rs:61]: better rename to vote_data_ordering_only

File: consensus/safety-rules/src/safety_rules.rs [hunk 5]
```diff
@@ -57,8 +58,9 @@ impl SafetyRules {
         persistent_storage: PersistentSafetyStorage,
         verify_vote_proposal_signature: bool,
         export_consensus_key: bool,
+        decoupled_execution: bool,
     ) -> Self {
-        let execution_public_key = if verify_vote_proposal_signature {
+        let execution_public_key = if verify_vote_proposal_signature && !decoupled_execution {
```
Comment (zekun000) [inline consensus/safety-rules/src/safety_rules.rs:63]: we don't need this when we already have the condition checked later?

File: consensus/consensus-types/src/vote_proposal.rs (hunk 4 above)
Comment (yuxiamit) [reply consensus/consensus-types/src/vote_proposal.rs:61]: resolved

**Review by zekun000 — COMMENTED** [review]
forget to mention, we should add at least one test for decouple_execution = true.

File: consensus/safety-rules/src/tests/suite.rs [hunk 6]
```diff
@@ -56,7 +56,7 @@ pub type Callback = Box<
     ),
 >;
 
-pub fn run_test_suite(safety_rules: &Callback) {
+pub fn run_test_suite(safety_rules: &Callback, decoupled_execution: bool) {
```
Comment (zekun000) [inline consensus/safety-rules/src/tests/suite.rs:59]: hmm, does test_bad_execution_output work with decoupled_execution = true?

Comment (yuxiamit) [reply consensus/safety-rules/src/tests/suite.rs:59]: I tried it locally and I didn't see a failure..?

File: consensus/safety-rules/src/tests/networking.rs [hunk 7]
```diff
@@ -6,15 +6,23 @@ use diem_types::validator_signer::ValidatorSigner;
...
 #[test]
 fn test_reconnect() {
-    let signer = ValidatorSigner::from_int(0);
-    let storage = test_utils::test_storage(&signer);
-    // test value for network timeout, in milliseconds.
-    let network_timeout = 5_000;
-    let safety_rules_manager =
-        SafetyRulesManager::new_thread(storage, false, false, network_timeout);
+    let boolean_values = [false, true];
+    for decoupled_execution in &boolean_values {
```
Comment (zekun000) [inline consensus/safety-rules/src/tests/networking.rs:10]: we probably can skip changing this one, seems irrelevant

**Review by zekun000 — APPROVED** [review]
